### DISCUSSION TITLE
Does brew have a technique to link a binary that still has access to its relative working directory in the app bundle?

### DISCUSSION THREAD

**Discussion by dannflor** [Homebrew/discussions#5012] [Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.1-18-gdbb800b
ORIGIN: https://github.com/Homebrew/brew
HEAD: dbb800b7d34b36c976d94c176decb2847de7d92a
Last commit: 30 hours ago
Core tap JSON: 28 Dec 06:05 UTC
Core cask tap JSON: 28 Dec 06:05 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 12
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: 15.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1
```

### Description of issue

I just merged https://github.com/Homebrew/homebrew-cask/pull/163355 which was meant to allow one to invoke `godot-mono` from the command line similar to how `godot` could be. I have now realized that maybe the reason no one else thought to add a link to the binary was that it searches the working directory for .NET assemblies and fails because they aren't in `/opt/homebrew/bin`. I'm sure this has been run into before. Does brew have a clever way to keep the working directory set to the App Bundle, or do we just not link when they depend on any resources there?

### Relevant casks

godot, godot-mono

**Comment by SMillerDev** [comment]
Try using https://docs.brew.sh/Formula-Cookbook#file-level-operations, mostly `write_exec_script`

**Reply by dannflor** [reply]
I tried `bin.write_exec_script "#{appdir}/Godot_mono.app/Contents/MacOS/Godot"` and it gave me `Error: Unexpected method 'bin' called on Cask godot-mono.`. If I try it without `bin`, it just says `Unexpected method 'write_exec_script'` instead. I'm using `brew install --cask --force`. Am I doing something wrong?

**Reply by SMillerDev** [reply]
Yeah, that doesn't work in casks. In casks the assumption is that upstream can and will provide a working distribution of the software that the cask only has to install.

**Reply by dannflor** [reply]
I think you might be misunderstanding me. The software works just fine when you, for instance, double-click it. I would just like a link to the binary to be available in /opt/homebrew/bin, something the `godot` cask was able to pull off just fine because it doesn't depend on any relatively pathed directories in the app bundle. I feel like this has to be a pretty common need, right? Don't apps bundle necessary resources all the time? Is it just impossible to use the `binary` function except when the executable is completely dependency-free?

**Reply by SMillerDev** [reply]
> Is it just impossible to use the binary function except when the executable is completely dependency-free?

It doesn't have to be dependency free, but it can't depend on the calling location to be the location of the binary, since that isn't always the case. This seems to be a bug upstream that needs to be resolved.

**Reply by dannflor** [reply, marked as answer]
I opened [this issue](https://github.com/godotengine/godot/issues/86625#issuecomment-1872604638) and got directed to [this technique](https://github.com/Homebrew/homebrew-cask/blob/e867f38d8e8a51c139fab025196a7ff84970e639/Casks/b/blender.rb#L22-L34), which appears to be the accepted way of duplicating `write_exec_script`'s behavior on a cask.
